lib/Entry.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module Entry
+  ( dateModified
+  , Entry, EntryConvertible (..), newEntry
+  , path, masterField, fields
+  , Field (..), FieldKey (..)
+  , newField, newFieldKey, getFieldKey, newEntryTag, getEntryTag
+  , private, value, tags
+  )
+where
+
+
+import qualified Data.Text as T
+import qualified Data.HashMap.Strict as HS
+
+import Control.Lens
+import qualified Data.Aeson.Types as A
+import Data.Hashable (Hashable)
+import GHC.Generics (Generic)
+import Data.Time (UTCTime)
+import Language.Haskell.TH.Lens (_CaseE)
+
+type DateTime = UTCTime
+
+newtype FieldKey = FieldKey T.Text
+  deriving (Generic, Show, Eq)
+
+instance Hashable FieldKey
+
+instance A.ToJSON FieldKey where
+instance A.ToJSONKey FieldKey where
+instance A.FromJSON FieldKey where
+instance A.FromJSONKey FieldKey where
+
+newFieldKey :: T.Text -> Maybe FieldKey
+newFieldKey t =
```

Comment:
We should also check that the input text has at least 1 character.

And because now we have two possible reasons for `newFieldKey` to fail, it should return an `Either Text FieldKey`, where `Text` contains the reason why it failed.

This also goes for `newEntryTag`